dra-evolution/pkg/api/claim_types.go

```diff
@@ -358,7 +282,7 @@ type ResourceRequestDetail struct {
 	// - 0 <= x <= maximum (up to a certain number, with zero instances acceptable)
 	// - minimum <= 0 <= maximum (within a certain range)
 	// +optional
-	Count *CountDetails `json:"count,omitempty"`
+	Count *IntRange `json:"count,omitempty"`
```

Comment:
Based on the discussion today, all we will support in 1.31 is "exact count". Does the user have to specify min: 5, max: 5 to get 5? That's not pleasant.

Crazy thought. What if we come up with a new scalar type that is an int range. So, the yaml looks like `count: 5` to get exactly 5, but later we allow `count: 1-4` for anywhere from one to four, or `count: 2,4` for either two or four. Tim is going punch me.

Kevin basically has this already for the capacity model; but this would put it in the user's view...

Reply:
Seems reasonable to me...

Reply:
I'll defer to @thockin here for the API review side of such a proposal.

Just one remark: such an IntRange type must take a string and store exactly that string in the apiserver. "Normalizing" it (`05-08` -> `5-8`) like the current IntRange implementation does is a big no-no and was called out as being a problem.

We also have to mindful of how we role this out. If we validate that string as "must be a single number" in 1.31 and in 1.32 allow it to be a range, then downgrading renders objects invalid. :boom:

In that sense, "all we will support in 1.31 is exact count" doesn't really help us, unless we interpret it as "we allow a range, but only implement a single value".

Reply:
I noted in a different comment (maybe it was in the doc not the PR) that since we have multiple requests per claim, and we have constraints across requests, we don't need to do multiple device per-request in 1.31. "Exact count of n" is equivalent to listing the same request n times. At least in the vast majority of cases (I can imagine cases where you have different constraints across different sets of devices, but I don't know of any real world use cases for that that we need to hit in 1.31).

That would mean we can eliminate this for now...

Reply:
If we leave out the `Count` field (or something like it) now, we cannot add it later because an old scheduler will not see the new field and incorrectly allocate a single device even though the user asked for several.

Repeating requests with identical requirements works for a fixed amount of devices, but not for the "give me management access to all devices" use case.

Regarding simplifying the "count" support to a single value instead of a range: I suspect that this will not really help reduce the amount of work. We still need a for loop and array somewhere, and validating the field becomes harder.

Reply:
I am not buying the "old schedulers won't know about a field and so will do something when they shouldn't". I don't think "doing nothing" is really great either. If a new feature adds the ability to constrain things in a way we did not allow in an old scheduler, is it really better to *simply ignore a pod and not schedule it*, than it is to schedule it according to the old rules?

We cannot build in stubs for everything we might ever need; there are lots of things we aren't thinking of right now. Users cannot and should not expect schedulers to be smart enough to magically know not to do something when new functionality has been added of which they are unaware.

Users have some responsibility to not utilize a new feature until it is fully rolled out. They also should not expect things that use the new feature will work as expected if they roll back.

I think we're distorting the API in really painful ways for "old schedulers won't know about it", but I think that is just not a real issue here. If we didn't have that assumption, we could probably cut out several one-ofs and fields, and create a tighter, simpler set of types that we can more easily reason about and explain to people.

Reply:
> If a new feature adds the ability to constrain things in a way we did not allow in an old scheduler, is it really better to simply ignore a pod and not schedule it, than it is to schedule it according to the old rules?

Suppose an app asks for 5 GPUs and wants to use one in each of its containers. It gets one GPU instead from an old scheduler and all containers share access to it, because the request index also gets ignored. Anything could happen when the app tries to run. Perhaps it fails to start but what if it instead runs and corrupts data?

Perhaps I am overthinking this. If @thockin as an API reviewer thinks that adding a new count field later is fine according to the [compatibility rules](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility) then I am okay with taking this out for now.

Reply:
We HAVE to assume that we can add fields over time, and that we can expand validation, given the right process.

> Tim is going punch me.

Why so violent?  But the thought did cross my mind.  I don't like string-encoding this, but I would be happy to argue about it later

Reply:
Let me clarify our options:
- we include it now in 1.31
- we don't include it in the KEP for 1.31, but sneak it in for 1.31 as a late update after all the other things are done
- we add it in 1.32 and immediately promote to beta
- we put it under a new feature gate ("DynamicResourceAllocationDeviceCounter") in 1.32 or later

What we cannot do is to add it to a beta API without a feature gate and just a comment saying "beware, this field is being ignored when some of your components in the cluster are too old" - that is what I meant earlier.

What I was trying with one-ofs was to avoid the need for feature gates. The new feature wouldn't work when older components are involved, but it would "fail safe".

Reply:
Without the one-ofs, we can add the fields that get ignored, but in a misconfigured cluster (feature gate enabled inconsistently) or with version skew it will not "fail safe" (see my example about asking for 5 GPUs and then running with 1).

This is only okay because we explicitly warned the user. But that doesn't make it nice.

Reply:
We decided to not use one-ofs to simplify the current API. The drawback will be that current clients may not "fail safely" when we add alternative fields in the future.